drivers/flash/nrf_qspi_nor: add support for enter 4-byte addressing mode

nvlsianpu opened the pull request:
Added basic support for enter 4-byte addressing command.
Patch  supports command 0xB7 (Enter 4-Byte Address Mode),
with or without preceding WREN.

Similar as for SPI-NOR the `enter-4byte-addr` property of
memory node is used or describing how to Enter 4-Byte Addressing
mode.
Worth to notice that along with that property the `address-size-32`
property is expected as it switch the driver to use 4-byte addressing
in operations.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

~~not tested on hardware yet~~
was tested using MX25L51245G

nvlsianpu committed:
drivers/flash/nrf_qspi_nor: add support for enter 4-byte addressing mode

Added basic support for enter 4-byte addressing command.
Patch  supports command 0xB7 (Enter 4-Byte Address Mode),
with or without preceding WREN.

Similar as for SPI-NOR the `enter-4byte-addr` property of
memory node is used or describing how to Enter 4-Byte Addressing
mode.
Worth to notice that along with that property the `address-size-32`
property is expected as it switch the driver to use 4-byte addressing
in operations.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

de-nordic commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -564,7 +576,20 @@ static int qspi_nrfx_configure(const struct device *dev)
 	}
 
-	return 0;
+	if (INST_0_4BA != 0) {
```
Maybe we should add LOG_DBG that would be stating that we got into 4-byte address mode.

nvlsianpu replied:
done

de-nordic commented:
Shouldn't we also address usage of sfdp-bfp in DTS? The enter_4byte_addr, as the binding states, is less important than sfdp bpfs, which stores the same info.

nvlsianpu commented:
> Shouldn't we also address usage of sfdp-bfp in DTS? The enter_4byte_addr, as the binding states, is less important than sfdp bpfs, which stores the same info.

This driver doesn't support DTS bundled sfd-bfp. I think runtime processing of these build time records is not something I should apply in this PR. I was looking into spi-nor driver: It looks like support of that possibly results in implementation of features unused in given application.

Anyway, adding  (runtime) read of DW16 of SFDP BFP make sense to me here.

nvlsianpu commented:
@de-nordic After I have been looking into spi_nor driver I conclude that the read and the runtime processing of SFDP might be implemented as there. Best approach would be extraction of common code from spi_nor.c (at last commonality of functionality of `spi_nor_process_sfdp()` function) and re-usage in both drivers. Such a task would be over what I want to introduce here.

de-nordic commented:
> @de-nordic After I have been looking into spi_nor driver I conclude that the read and the runtime processing of SFDP might be implemented as there. Best approach would be extraction of common code from spi_nor.c (at last commonality of functionality of `spi_nor_process_sfdp()` function) and re-usage in both drivers. Such a task would be over what I want to introduce here. #41819

Great. So I leave adding the LOG_DBG line up to you. I have looked at the code and it looks OK.
In the future we should probably add some defines to cover the bit meanings in the BFP bytes, but that will be a lot of work.

de-nordic reviewed (dismissed):
Looks OK. Thanks.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -564,7 +576,26 @@ static int qspi_nrfx_configure(const struct device *dev)
 
-	return 0;
+	if (INST_0_4BA != 0) {
+		if (INST_0_4BA & 0x02) {
+			qspi_nor_write_protection_set(dev, false);
+		}
+
+		struct qspi_cmd cmd = {
+			.op_code = SPI_NOR_CMD_4BA,
+			.rx_buf = NULL,
+			.tx_buf = NULL,
+		};
+		ret = qspi_send_cmd(dev, &cmd, false);
```
How about just:
```
		struct qspi_cmd cmd = {
			.op_code = SPI_NOR_CMD_4BA,
		};
		ret = qspi_send_cmd(dev, &cmd, (INST_0_4BA & 0x02));
```
?

nvlsianpu replied:
and right, rest of struct members will be zeroed automatically.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -564,7 +576,26 @@ static int qspi_nrfx_configure(const struct device *dev)
+		};
+		ret = qspi_send_cmd(dev, &cmd, false);
+
+		if (ret < 0) {
+			LOG_DBG("E4BA cmd issued.");
+		} else {
+			LOG_ERR("E4BA cmd issue failed: %d.", ret);
+		}
```
`ret < 0` means that an error occurred, so this should be the other way around.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -79,6 +79,18 @@ BUILD_ASSERT(((INST_0_QER == JESD216_DW15_QER_NONE)
 	      || (INST_0_QER == JESD216_DW15_QER_S1B6)),
 	     "Driver only supports NONE or S1B6 for quad-enable-requirements");
 
+#if DT_INST_NODE_HAS_PROP(0, enter_4byte_addr)
+#define INST_0_4BA DT_INST_PROP(0, enter_4byte_addr)
```
```suggestion
#define INST_0_4BA DT_INST_PROP_OR(0, enter_4byte_addr, 0)
```

nvlsianpu commented:
@richesonk I'm going to test this finally.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -79,6 +79,16 @@ BUILD_ASSERT(((INST_0_QER == JESD216_DW15_QER_NONE)
+#if (INST_0_4BA != 0)
+BUILD_ASSERT(((INST_0_4BA & 0x03) != 0),
+	     "Driver only supports command (0xB7) for entering 4 byte addressing mode");
+BUILD_ASSERT((DT_INST_PROP(0, address_size_32) == NRF_QSPI_ADDRMODE_32BIT),
```
```suggestion
BUILD_ASSERT(DT_INST_PROP(0, address_size_32),
```
This is a boolean property, so it should not be compared with the value of `NRF_QSPI_ADDRMODE_32BIT` (which by chance evaluates to 1, hence it works in the current form).

nvlsianpu committed:
tests/driver/flash: overlay for MX25L51245G

Added overlay with DTS of MX25L51245G qspi flash
memory. DTS configures qspi clock to 2 MHz which is supported
by nRF52840.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

dastarling commented:
I also need this change for an active development project soon.  I hope it makes the next NCS release. Thanks

nvlsianpu commented:
@de-nordic @anangl Can you revisit?

anangl commented on tests/drivers/flash/boards/nrf52840dk_mx25l51245g.overlay:
```diff
@@ -0,0 +1,37 @@
+
+/delete-node/ &mx25r64;
+
+&qspi {
+	sck-pin = <30>;
+	io-pins = <29>, <28>, <04>, <03>;
+	csn-pins = <31>;
```
This board uses PINCTRL now, so you need to rebase and then modify `pinctrl-*` properties instead.

```suggestion
&pinctrl {
	qspi_default: qspi_default {
		group1 {
			psels = <NRF_PSEL(QSPI_SCK, 0, 30)>,
				<NRF_PSEL(QSPI_IO0, 0, 29)>,
				<NRF_PSEL(QSPI_IO1, 0, 28)>,
				<NRF_PSEL(QSPI_IO2, 0, 04)>,
				<NRF_PSEL(QSPI_IO3, 0, 03)>,
				<NRF_PSEL(QSPI_CSN, 0, 31)>;
		};
	};

	qspi_sleep: qspi_sleep {
		group1 {
			psels = <NRF_PSEL(QSPI_SCK, 0, 30)>,
				<NRF_PSEL(QSPI_IO0, 0, 29)>,
				<NRF_PSEL(QSPI_IO1, 0, 28)>,
				<NRF_PSEL(QSPI_IO2, 0, 04)>,
				<NRF_PSEL(QSPI_IO3, 0, 03)>,
				<NRF_PSEL(QSPI_CSN, 0, 31)>;
			low-power-enable;
		};
	};
};

&qspi {
	pinctrl-0 = <&qspi_default_alt>;
	pinctrl-1 = <&qspi_sleep_alt>;
	pinctrl-names = "default", "sleep";
```

nvlsianpu replied:
OK

nvlsianpu commented on tests/drivers/flash/boards/nrf52840dk_mx25l51245g.overlay:
```diff
@@ -0,0 +1,37 @@
+
```
ok

nvlsianpu commented on tests/drivers/flash/testcase.yaml:
```diff
@@ -5,6 +5,13 @@ tests:
+    platform_allow: nrf52840dk_nrf52840
+    tags: flash nrf52 nrf_qspi_fash
+    extra_args: OVERLAY_CONFIG=boards/nrf52840_flash_qspi.conf
+      DTC_OVERLAY_FILE=./boards/nrf52840dk_mx25l51245g.overlay
```
'./' is not needed.
